pkg/daemon/daemon.go

```diff
@@ -398,54 +435,65 @@ func (dn *Daemon) nodeStateChangeHandler(old, new interface{}) {
 		// Apply generic_plugin last
 		err = dn.LoadedPlugins[GenericPlugin].Apply()
 		if err != nil {
-			glog.Errorf("nodeStateChangeHandler(): generic_plugin fail to apply: %v", err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): generic_plugin fail to apply: %v", err)
+			return err
 		}
 	}
 
 	if reqReboot {
-		glog.Info("nodeStateChangeHandler(): reboot node")
-		go rebootNode()
-		return
+		glog.Info("nodeStateSyncHandler(): reboot node")
+		rebootNode()
+		return nil
 	}
 
 	// restart device plugin pod
-	if reqDrain || (latestState.Spec.DpConfigVersion != oldState.Spec.DpConfigVersion) {
-		glog.Info("nodeStateChangeHandler(): restart device plugin pod")
+	if reqDrain || (dn.nodeState.GetName() != "" && latestState.Spec.DpConfigVersion != dn.nodeState.Spec.DpConfigVersion) {
+		glog.Info("nodeStateSyncHandler(): restart device plugin pod")
 		if err := dn.restartDevicePluginPod(); err != nil {
-			glog.Errorf("nodeStateChangeHandler(): fail to restart device plugin pod: %v", err)
-			dn.exitCh <- err
-			return
+			glog.Errorf("nodeStateSyncHandler(): fail to restart device plugin pod: %v", err)
+			return err
 		}
 	}
 
 	if anno, ok := dn.node.Annotations[annoKey]; ok && anno == annoDraining {
-		dn.completeDrain()
+		if err := dn.completeDrain(); err != nil {
+			glog.Errorf("drainNode(): failed to complete draining: %v", err)
+			return err
+		}
 	} else if !ok {
 		if err := dn.annotateNode(dn.name, annoIdle); err != nil {
 			glog.Errorf("drainNode(): failed to annotate node: %v", err)
-			dn.exitCh <- err
-			return
+			return err
```

Comment:
In which case, will this `else if` be executed?

Reply:
Before the node is annotated.

Reply:
Every node will first be annotated as Draining, then be marked as Idle, how would the first time annotating happen?

Reply:
When the operator is deployed, the node will be annotated to Idle for the first time.

Reply:
Ok, would it make sense to move the annotating action into completeDrain? this will make annotating associated with draining process, e.g annotating `draining` in `drainNode` and annotating `idle` in `completeDrain`.

Reply:
> Ok, would it make sense to move the annotating action into completeDrain? this will make annotating associated with draining process, e.g annotating `draining` in `drainNode` and annotating `idle` in `completeDrain`.

This is the current implementable already. The `annotateNode` here is just for setting the annotation to the initial value.